tools/health-monitor/monitor/file_cache/cache_monitor.go

```diff
@@ -0,0 +1,311 @@
+		return err
+	}
+
+	// set recursive watcher on file cache directory
+	if err := w.AddRecursive(fc.tmpPath); err != nil {
+		log.Err("cache_monitor::cacheWatcher : [%v]", err)
+		return err
+	}
+
+	// list of all of the files and folders currently being watched
+	for path := range w.WatchedFiles() {
+		log.Debug("cache_monitor::cacheWatcher : Watching %v", path)
+	}
+
+	// Start the watching process - it'll check for changes every 100ms
+	if err := w.Start(time.Millisecond * 100); err != nil {
+		log.Err("cache_monitor::cacheWatcher : [%v]", err)
+		return err
+	}
+
+	return nil
+}
+
+func (fc *FileCache) createEvent(event *watcher.Event) {
+	if !event.IsDir() {
+		delete(fc.cacheObj.fileRemovedMap, event.Path)
+		fc.cacheObj.fileCreatedMap[event.Path] = event.Size()
+		fc.cacheObj.cacheSize += event.Size()
+		fc.cacheObj.cacheConsumed = (float64)(fc.cacheObj.cacheSize*100) / (fc.maxSizeMB * common.MbToBytes)
+	}
+
+	e := fc.getCacheEventObj(event)
+	if !event.IsDir() {
+		e.Value[fileSize] = strconv.FormatInt(event.Size(), 10)
+	}
+
+	log.Debug("cache_monitor::createEvent : %v", e)
+	fc.ExportStats(time.Now().Format(time.RFC3339), e)
+}
+
+func (fc *FileCache) removeEvent(event *watcher.Event) {
+	if !event.IsDir() {
+		delete(fc.cacheObj.fileCreatedMap, event.Path)
+		fc.cacheObj.fileRemovedMap[event.Path] = event.Size()
+		fc.cacheObj.cacheSize = int64(math.Max(0, float64(fc.cacheObj.cacheSize-event.Size())))
+		fc.cacheObj.cacheConsumed = (float64)(fc.cacheObj.cacheSize*100) / (fc.maxSizeMB * common.MbToBytes)
+	}
+
+	e := fc.getCacheEventObj(event)
+	if !event.IsDir() {
+		e.Value[fileSize] = strconv.FormatInt(event.Size(), 10)
+	}
+
+	log.Debug("cache_monitor::removeEvent : %v", e)
+	fc.ExportStats(time.Now().Format(time.RFC3339), e)
+}
+
+func (fc *FileCache) chmodEvent(event *watcher.Event) {
+	if !event.IsDir() {
+		delete(fc.cacheObj.fileRemovedMap, event.Path)
+		fileSize := fc.cacheObj.fileCreatedMap[event.Path]
+
+		if fileSize == event.Size() {
+			return
+		}
+
+		fc.cacheObj.cacheSize += event.Size() - fileSize
+		fc.cacheObj.fileCreatedMap[event.Path] = event.Size()
+		fc.cacheObj.cacheConsumed = (float64)(fc.cacheObj.cacheSize*100) / (fc.maxSizeMB * common.MbToBytes)
+	}
+
+	e := fc.getCacheEventObj(event)
+	e.Value[mode] = event.Mode().String()
+
+	log.Debug("cache_monitor::chmodEvent : %v", e)
+	fc.ExportStats(time.Now().Format(time.RFC3339), e)
+}
+
+func (fc *FileCache) writeEvent(event *watcher.Event) {
+	if !event.IsDir() {
+		delete(fc.cacheObj.fileRemovedMap, event.Path)
+		fileSize := fc.cacheObj.fileCreatedMap[event.Path]
+
+		if fileSize == event.Size() {
+			return
+		}
+
+		fc.cacheObj.cacheSize += event.Size() - fileSize
+		fc.cacheObj.fileCreatedMap[event.Path] = event.Size()
+		fc.cacheObj.cacheConsumed = (float64)(fc.cacheObj.cacheSize*100) / (fc.maxSizeMB * common.MbToBytes)
+	} else {
+		return
+	}
+
+	// not logging write events since there are multiple events in case of uploads
```

Comment:
so we will never log this and just let the blobfuse stats take care of the file download progress?

Reply:
That's right. For large files upload/download, there can lots of write events. So, I am neither logging nor exporting the write events. The upload/download progress part is taken care by the blobfuse binary where it sends an event to the monitor binary for every 100MB of data transferred.
The reason why we have the writeEvent() method is because for upload operation we don't get the complete file size during the create event. So, writeEvent() method takes care of updating the file size and cache consumption stats.